### PR TITLE
Added a polarization module to sympy.physics.optics

### LINKED ISSUE #17679: Add Jones calculus module to sympy.physics.optics
Let's add a module that provides Jones matrices for half, quarter, and generic phase retarders. Also linear polarizers and polarizing prisms.

### PR DESCRIPTION
**Pull request by oscarlazoarjona**
#### Fixes #17679

* physics.optics
  * added module `polarization`.

### CODE REVIEW

File: sympy/physics/optics/polarization.py [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+details and examples.
+
+Credits and Copyright
+~~~~~~~~~~~~~~~~~~~~~
+
+AUTHOR:
+
+- Oscar Gerardo Lazo Arjona (2017-06-18): added Wigner D matrices
+
+Copyright (C) 2019 Oscar Gerardo Lazo Arjona <algebraicamente@gmail.com>
```
Comment (asmeurer) [inline sympy/physics/optics/polarization.py]: Remove these lines. The authorship of SymPy is reflected in the git history and in the AUTHORS file (which is generated automatically).

**Comment by asmeurer** [comment]
There may be issues with using Unicode in the doctests. I don't believe we do it anywhere else. We should stick with ASCII for now. Actually I believe most of the examples will be readable just fine with the non-pprint output, because the str form of Matrix prints with newlines.

**Comment by asmeurer** [comment]
Some comments:

- This needs tests
- This needs to be added to `sympy/physics/optics/__init__.py`
- This should be added to the Sphinx docs (in doc/)
- Instead of INPUT and OUTPUT in the tests use the "Parameters" and "Returns" headers. See the style guide at https://github.com/sympy/sympy/pull/17715, which will be merged soon, or look at the numpydoc guide for how to format it.

I cannot comment on the code itself as I am not familiar with the physics.

File: sympy/physics/optics/polarization.py (hunk 1 above)
Comment (oscarlazoarjona) [reply sympy/physics/optics/polarization.py]: Ok. There are similar lines in my previous contribution, though. I guess those should also be removed. Is that done in a separate PR?

**Comment by oscarlazoarjona** [comment]
> Some comments:
> 
>     * This needs tests
> 
>     * This needs to be added to `sympy/physics/optics/__init__.py`
> 
>     * This should be added to the Sphinx docs (in doc/)
> 
>     * Instead of INPUT and OUTPUT in the tests use the "Parameters" and "Returns" headers. See the style guide at #17715, which will be merged soon, or look at the numpydoc guide for how to format it.
> 
> 
> I cannot comment on the code itself as I am not familiar with the physics.

Wilco.

File: sympy/physics/optics/polarization.py (hunk 1 above)
Comment (asmeurer) [reply sympy/physics/optics/polarization.py]: Yes, do that in a separate PR.

**Comment by oscarlazoarjona** [comment]
> There may be issues with using Unicode in the doctests. I don't believe we do it anywhere else. We should stick with ASCII for now. Actually I believe most of the examples will be readable just fine with the non-pprint output, because the str form of Matrix prints with newlines.

Similar to my previous anwer, I have already submitted code with unicode doctests. Should that too be reverted to ASCII? The doctests passed fine on Travis for those examples.

**Comment by asmeurer** [comment]
Let's open an issue about it, but you may want to. The biggest issue is that people without the proper fonts (e.g., on mobile devices) won't be able to read the online documentation.

File: sympy/physics/optics/polarization.py [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+
+from sympy import sin, cos, exp, I, pi, sqrt, Matrix, Abs, re, im, simplify
+from sympy.physics.quantum import TensorProduct
+
+
+def jones_vector(psi, chi):
+    u"""A Jones vector corresponding to a polarization ellipse with `psi` tilt,
+    and `chi` circularity.
+
+    INPUT:
```
Comment (jksuom) [inline sympy/physics/optics/polarization.py]: Please use standard section headings like "Parameters" as explained [here](https://numpydoc.readthedocs.io/en/latest/format.html#sections).

Comment (oscarlazoarjona) [reply sympy/physics/optics/polarization.py]: Wilco.

**Comment by oscarlazoarjona** [comment]
@jksuom @asmeurer, could you point me to the error in [this CI test](https://travis-ci.org/sympy/sympy/jobs/604320205?utm_medium=notification&utm_source=github_status) I don't understand what the problem is.

File: sympy/physics/optics/polarization.py (hunk 1 above)
Comment (oscarlazoarjona) [reply sympy/physics/optics/polarization.py]: I checked the AUTHORS file, and my name isn't there. Also, running bin/authors_update.py doesn't seem to update the file. How do I add my name?

**Comment by jksuom** [comment]
It seems that the flake8 test is currently always failing, but that is an allowed failure. It is the [Sphinx test](https://travis-ci.org/sympy/sympy/jobs/604320193?utm_medium=notification&utm_source=github_status) that will make the Travis checks fail.

File: sympy/physics/optics/polarization.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+- Jones vectors.
+
+- Stokes vectors.
+
+- Jones matrices.
+
+- Mueller matrices.
+
+Please see the description of the individual functions for further
+details and examples.
```
Comment (oscarbenjamin) [inline sympy/physics/optics/polarization.py]: The new documentation guide
https://docs.sympy.org/dev/documentation-style-guide.html
says
```
The module level docstring should discuss the purpose and scope of the module, and give a high-level example of how to use the functions or classes in the module
```
I expected there to be more guidance there about module level docstrings but when I looked this is what I found.

File: sympy/physics/optics/polarization.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/physics/optics/polarization.py]: I'm not sure how the authors file is updated but I guess that you should be there given previous contributions...

File: sympy/physics/optics/polarization.py (hunk 3 above)
Comment (oscarlazoarjona) [reply sympy/physics/optics/polarization.py]: I just added further description of the module and examples.

File: sympy/physics/optics/polarization.py (hunk 1 above)
Comment (asmeurer) [reply sympy/physics/optics/polarization.py]: We update the AUTHORS periodically, and will at least do so before making a release. 

You can update it yourself it you want (but please do a separate PR). It requires running that script and then fixing any duplicates.

**Comment by oscarbenjamin** [comment]
This looks good

**Pull request merged**